## Table of contents
## Step 1-3: Importing data files and libraries
## Step 4: Wrangle the data
## Step 5: Complete the fundamental data quality and consistency checks 
## Step 6: Combine your customer data with the rest of your prepared Instacart data

## Step 1-3: Importing data files and libraries

In [3]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [4]:
# Importing customer data file

path = r'C:\Users\Lenovo\Documents\Instacart Basket Analysis'
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'),index_col = False)

## Step 4: Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [9]:
# Checking the column headers of the df

df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [10]:
# Checking the dimensions of the df

df_customers.shape

(206209, 10)

In [5]:
# Renaming columns to be more more intuitive: changing Surnam to Last Name

df_customers.rename(columns = {'Surnam': 'Last Name'}, inplace = True)

In [6]:
# Renaming columns to be more intuitive: changing n_dependants to Number of Dependants

df_customers.rename(columns = {'n_dependants': 'Number of Dependants'}, inplace = True)

In [7]:
# Renaming columns to be more intuitive: changing fam_status to Marital Status

df_customers.rename(columns = {'fam_status': 'Marital Status'}, inplace = True)

In [8]:
# Checking the new column headers

df_customers.head()

user_id First Name Last Name  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number of Dependants Marital Status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

I'm not sure which columns I will need for further analysis, so for the moment I will leave them as they are.

## Step 5: Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data

In [15]:
# Checking data types

df_customers.dtypes

user_id                  int64
First Name              object
Last Name               object
Gender                  object
STATE                   object
Age                      int64
date_joined             object
Number of Dependants     int64
Marital Status          object
income                   int64
dtype: object

In [16]:
# While being a number, user_id contains only categorical information, so it could be 
# changed to a string variable.

df_customers['user_id'] = df_customers['user_id'].astype('str')

In [17]:
# Checking the output

df_customers['user_id'].dtypes

dtype('O')

In [18]:
# Checking the descriptive statistics of the numerical columns

df_customers.describe()

Age  Number of Dependants         income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

In [19]:
# Checking for missing values

df_customers.isnull().sum()

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
STATE                       0
Age                         0
date_joined                 0
Number of Dependants        0
Marital Status              0
income                      0
dtype: int64

There are 11259 missing values in the First Name column. I will fill the empty cells with NaN values since there is no way that I can imput or guess the missing values. 

In [9]:
# Filling the empty cells with NaN values

df_customers['First Name'].fillna('NaN', inplace=True)

In [25]:
# Checking for mixed-type data columns

for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)
  else:
    print ('No mixed dtypes', col)

No mixed dtypes user_id
No mixed dtypes First Name
No mixed dtypes Last Name
No mixed dtypes Gender
No mixed dtypes STATE
No mixed dtypes Age
No mixed dtypes date_joined
No mixed dtypes Number of Dependants
No mixed dtypes Marital Status
No mixed dtypes income


There are no mixed-type data columns. 

In [26]:
# Exporting data file

df_customers.to_pickle(os.path.join(path, '02 Data','Original Data', 'customers_2.pkl'))

In [4]:
# Importing data file

path = r'C:\Users\Lenovo\Documents\Instacart Basket Analysis'
df_customers = pd.read_pickle(os.path.join(path, '02 Data', 'Original Data', 'customers_2.pkl'))

In [10]:
# Checking for duplicate records by creating a new dataframe that contains duplicate records

df_dups = df_customers[df_customers.duplicated()]

In [12]:
df_dups.head()

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, STATE, Age, date_joined, Number of Dependants, Marital Status, income]
Index: []

There are no duplicate records in the dataframe.

## Step 6: Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [12]:
# Importing the ords_prods_merged data file

df_ords_prods_merged = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_4.csv'), index_col = False)

In [13]:
# Checking the datafile

df_ords_prods_merged.shape

(32404859, 25)

In [14]:
df_ords_prods_merged.columns

Index(['Unnamed: 0', 'order_id', 'user_id', 'order_number',
       'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', '_merge', 'price_label',
       'busiest_day', 'busiest_day_2', 'busiest_hour_of_day', 'max_order',
       'loyalty_flag', 'avg_price', 'spending_flag', 'median',
       'ord_frequency_flag'],
      dtype='object')

In [15]:
# Dropping an unnecessary column

df_ords_prods_merged = df_ords_prods_merged.drop(columns = ['Unnamed: 0']) 

In [16]:
df_ords_prods_merged.shape

(32404859, 24)

In [17]:
# Checking the datatypes

df_ords_prods_merged.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
product_id                  int64
add_to_cart_order           int64
reordered                   int64
product_name               object
aisle_id                    int64
department_id               int64
prices                    float64
_merge                     object
price_label                object
busiest_day                object
busiest_day_2              object
busiest_hour_of_day        object
max_order                   int64
loyalty_flag               object
avg_price                 float64
spending_flag              object
median                    float64
ord_frequency_flag         object
dtype: object

In [ ]:
# Changing the datatype of the column user_id to string, so that it is the same datatype as user_id in the customers df

In [18]:
df_ords_prods_merged['user_id'] = df_ords_prods_merged['user_id'].astype('str')

In [19]:
df_ords_prods_merged['user_id'].dtypes

dtype('O')

In [20]:
# Merging the two dfs with user_id as key variable

df_combined = df_ords_prods_merged.merge(df_customers, on = 'user_id')

In [21]:
# Cleaning the column headings for consistency

df_combined.rename(columns = {'First Name' : 'first_name', 'Last Name' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'Number of Dependants' : 'number_of_dependants', 'Marital Status' : 'marital_status'}, inplace = True)

In [22]:
df_combined.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     0.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  ord_frequency_flag  first_name  last_name  gender  \
0         Soda  ...    Regular customer       Linda     Nguyen  Female   
1         Soda  ...    Regular customer       Linda     Nguyen  Female   
2         Soda  ...    Regular customer       Linda     Nguyen  Female   
3         Soda  ...    Regular customer       Linda     Nguyen  Female   
4         Soda  ...    Regular customer       Linda     Nguyen  Female   

     state age date_joined number_of_dependants  marital_status income  
0  Alabama  31   2/17/2019                    3         married  40423  
1  Alabama  31   2/17/2019                    3         married  40423  
2  Alabama  31   2/17/2019                    3         married  40423  
3  Alabama  31   2/17/2019                    3         married  40423  
4  Alabama  31   2/17/2019                    3         married  40423  

[5 rows x 33 columns]

In [23]:
# Exporting the datafile

df_combined.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'all_combined.pkl'))